src/effects/How to make an effect stateless.txt

```diff
@@ -0,0 +1,189 @@
+           {
+              // pass back the modified settings to the MessageBuffer
+
+              // TODO uncomment at last step
+              //EffectEcho::GetSettings(settings) = mSettings;
+           }
+           );
+
+           return true;
+        }
+
+
+        bool EffectEcho::Validator::UpdateUI()
+        {
+           // get the settings from the MessageBuffer and write them to our local copy
+           const auto& settings = mAccess.Get();
+
+           // TODO uncomment at last step
+           //mSettings = GetSettings(settings);
+
+           return true;
+        }
+
+    
+
+
+3) "Define the instance"
+
+    In header file, 
+    
+        add in public section:
+    
+            struct Instance;
+
+            std::shared_ptr<EffectInstance> MakeInstance(EffectSettings& settings) const;
+
+        Identify members that represent state - comment them out 
+        
+        remove methods: ProcessInitialize, ProcessFinalize, ProcessBlock
+
+    In cpp file,
+    
+        add:
+    
+        struct EffectEcho::Instance
+           : public PerTrackEffect::Instance
+           , public EffectInstanceWithBlockSize
+           , public EffectInstanceWithSampleRate
+
+        {
+           Instance(const PerTrackEffect& effect)
+              : PerTrackEffect::Instance{ effect }
+           {}
+
+           bool ProcessInitialize(EffectSettings& settings,
+              sampleCount totalLen, ChannelNames chanMap) override;
+
+           size_t ProcessBlock(EffectSettings& settings,
+              const float* const* inBlock, float* const* outBlock, size_t blockLen)  override;
+
+           bool ProcessFinalize(void) override;
+
+           << MEMBERS WHICH REPRESENT STATE which you commented out earlier >>
+        };
+
+
+        std::shared_ptr<EffectInstance>
+        EffectEcho::MakeInstance(EffectSettings&) const
+        {
+           return std::make_shared<Instance>(*this);
+        }
+        
+        
+    Make the 3 Effect::Process*** methods belong to the instance by: 
+    
+    - prepending ::Instance where needed
+
+    - adding this at the beginning (may not be needed in all of them)
+    
+        // temporary - in the final step this will be replaced by
+        // auto& echoSettings = GetSettings(settings);
+        //
+        auto& echoSettings = static_cast<const EffectEcho&>(mProcessor).mSettings;
+
+    
+
+
+4) "Real Statelessness"
+
+    In header file:
+    
+    - make the effect inherit  EffectWithSettings<EffectEchoSettings, PerTrackEffect>
+    
+    - remove the FetchParameters method
+    
+    - remove mSettings member
+
+
+    In cpp file:
+    
+    - go to processInitialize, ProcessBlock, ProcessFinalize; remove the downcast and enable the commented out GetSettings
+    
+    - go to Validator::ValidateUI & Validator::UpdateUI, remove the commenting out on the commented lines
+    
+    - add const to arg EffectEchoSettings& in the Validator constructor
```

Comment:
Or (as in LadspaEffect::Validator) you might need only access to get the settings to be copied:

```
   Validator(EffectUIClientInterface &effect,
      EffectSettingsAccess &access, double sampleRate, EffectType type)
      : EffectUIValidator{ effect, access }
/* other member initializations */
      // Copy settings
      , mSampleRate{ sampleRate }
      , mType{ type }
      , mSettings{ GetSettings(access.Get()) }
   {}
```